### PR TITLE
Matrix inversion

### PR DESCRIPTION
**Pull request by tschw**
Surprisingly `my3x3.getInverse` takes a 4x4 matrix which is rather odd; an exposed implementation detail of the normal matrix calculation. 3x3 inversion is useful, so this PR adds a real 3x3 inversion routine and inlines the previous code into the normal matrix calculation.

### CODE REVIEW

**Comment by WestLangley** [comment]
>Matrix3.getInverse takes a 4x4 matrix which is rather odd

It is rather odd. But you are changing the signature of the method, which will cause breakage. There would need to be some type checking and warning in case of misuse.

>3x3 inversion is useful

It is useful, but so is an eigenvector method -- and a host of other methods. This is not a math library, and the `Matrix3` class has been limited to methods required for the library to function. That is why it is the way it is. : - )

That being said, I can go either way on this one.

File: src/math/Matrix3.js [hunk 1]
```diff
@@ -152,25 +153,18 @@ THREE.Matrix3.prototype = {
...
 
-		// input: THREE.Matrix4
-		// ( based on http://code.google.com/p/webgl-mjs/ )
+		var me = matrix.elements,
+			te = this.elements,
 
-		var me = matrix.elements;
-		var te = this.elements;
+			n11 = me[ 0 ], n21 = me[ 1 ], n31 = me[ 2 ],
+			n12 = me[ 3 ], n22 = me[ 4 ], n32 = me[ 5 ],
```
Comment (WestLangley) [inline src/math/Matrix3.js]: Question: what is better? Nine new variables that need to be garbage-collected, or the repeated array accesses?

File: src/math/Matrix3.js [hunk 2]
```diff
@@ -230,13 +231,35 @@ THREE.Matrix3.prototype = {
...
+		te[ 6 ] =	me[ 9 ] * me[ 4 ] - me[ 5 ] * me[ 8 ];
+		te[ 7 ] = - me[ 9 ] * me[ 0 ] + me[ 1 ] * me[ 8 ];
+		te[ 8 ] =	me[ 5 ] * me[ 0 ] - me[ 1 ] * me[ 4 ];
+
+		var det = me[ 0 ] * te[ 0 ] + me[ 1 ] * te[ 3 ] + me[ 2 ] * te[ 6 ];
+
+		if ( det === 0 ) {
+
+			console.warn(
+					"THREE.Matrix3.getInverseOfTranslated(): can't invert matrix, determinant is 0" );
```
Comment (WestLangley) [inline src/math/Matrix3.js]: "THREE.Matrix3.getNormalMatrix(): determinant is 0"

**Comment by WestLangley** [comment]
What is going on with your `Matrix4.getInverse()` mods. Is the new code better by some measure?

File: src/math/Matrix3.js (hunk 1 above)
Comment (tschw) [inline src/math/Matrix3.js]: Local variables are ethereal and not subjected to garbage collection. They may hold new object (they don't in this case) in which case *the objects* get collected once they are no longer referenced anywhere.

**Comment by tschw** [comment]
@WestLangley 
> What is going on with your Matrix4.getInverse() mods? Is the new code better by some measure?

Code quality.

It's good practice to throw up before altering the state of an object. So is reading arrays linearly when you can. 

But since

> This is not a math library, and the Matrix3 class has been limited to methods required for the library to function.

we might as well just remove the `throw` altogether - don't see it being used anywhere.

**Comment by tschw** [comment]
@WestLangley 
> > Matrix3.getInverse takes a 4x4 matrix which is rather odd

> It is rather odd. But you are changing the signature of the method, which will cause breakage. There would need to be some type checking and warning in case of misuse.

I checked before making this PR that there are no references to this method within Three.js by manual inspection of all occurrences of the string "getInverse" in the code base.
Due to the odd semantics it's very unlikely to find it referenced somewhere in client code, but we never know... 

I added a check just in case. It's a separate commit so I can easily take it out again in case you have a change of heart. Just let me know.

**Comment by WestLangley** [comment]
>we might as well just remove the throw altogether - don't see it being used anywhere.

/ping @bhouston What is your opinion on this? Ref: https://github.com/mrdoob/three.js/commit/f52aa6303dbe7dcfb0d52e513e195b9befd1d7f3

**Comment by bhouston** [comment]
We use the throw, it is super useful.  I do not thinking hiding a failed inversion is good idea.  How would one check if it failed to invert if we removed the throw?  Checking for identity is not a robust check.  I am okay if we change getInverse to return a true/false boolean as to whether it worked, I could live with that.

That said, we patch Three.JS anyhow in Clara.io so you can remove it and we'll just patch it the way we want.

**Comment by bhouston** [comment]
I do think that being able to respond when something fails is an important part of robust code.  I think that there are very few places in the math library that can fail, but matrix inversion is one of them.  Other functions that can fail, such as Ray.intersectplane have clear failure paths, such as returning null: https://github.com/mrdoob/three.js/blob/master/src/math/Ray.js#L310  I would like to have a failure path for functions that potentially fail. 

See: https://en.wikipedia.org/wiki/Error_hiding

**Comment by tschw** [comment]
> We use the throw, it is super useful. I do not thinking hiding a failed inversion is good idea. How would one check if it failed to invert if we removed the throw? Checking for identity is not a robust check.

As a matter of my personal taste, I prefer an exception over a log message, code that crashes in the right spot over halfway working stuff, and I don't expect a division by zero to yield one. It's just the library seems to prefer the other code path for whatever reason.

Having users that type in matrices is of course a strong argument to be able to detect it. Otherwise inverting a singular matrix is just an internal error that simply should never happen in a finished program. 

BTW [`console.assert`](https://developer.mozilla.org/en-US/docs/Web/API/Console/assert) seems to exist in all browsers, so that one might be a better choice than `console.warn`, because it always prints a stack trace (but unlike in other programming languages does not throw an exception).

**Comment by WestLangley** [comment]
Still thinking about this...

This PR is about changing the arg to `Matrix3.getInverse()` from a `Matrix4` to `Matrix3`. It was originally written that way because it was a utility method used by `getNormalMatrix()`, which takes a `Matrix4`.

I do not think the change to `Matrix3.getInverse()` is necessary, really. Actually, I think I would prefer removing the method now, since this PR inlines `getNormalMatrix()`, and `getInverse()` is no longer needed by the library. (We will find out soon enough if users are using it).

@tschw You have inlined `getNormalMatrix()`. Previously, the code was simple:

    this.getInverse( m ).transpose();

which is familiar to anyone who understands the math... With this PR, the code is not so clear.

If you want to go the inline route, anyway, perhaps it would be wise to remove the `transpose()` call, and inline the entie method.

**Comment by tschw** [comment]
Another option is to add another generally useful method `fromMatrix4` and then go like
```javascript
getNormalMatrix: function ( m ) {

    return this.fromMatrix4( m ).getInverse( this ).transpose();

}
```

**Comment by tschw** [comment]
@WestLangley 
> I do not think the change to Matrix3.getInverse() is necessary, really. 

If the method is present, it should do the obvious thing, which is invert a 3x3 matrix or have a different name.

**Comment by WestLangley** [comment]
>Another option is to add another generally useful method fromMatrix4 and then go like

I like that -- as long as inverting in place works. That way you can keep your new `Matrix3.getInverse()` signature. : - )

To keep with the naming convention, maybe `Matrix3.setFromMatrix4()`.

**Comment by tschw** [comment]
@WestLangley 
> I like that

Cool, me too! So be it, then.

**Comment by tschw** [comment]
@WestLangley 
> That way you can keep your new Matrix3.getInverse() signature. : - )

Looking at it now, I should have written it like this right from the start. But then again, it's why there's a review :-).

The branch has been tested and works.

**Comment by WestLangley** [comment]
@tschw Sweet!

@bhouston Thank you for your input.

It looks like everybody is happy. Miracles do happen. : - )

**Comment by mrdoob** [comment]
Thanks guys!

**Pull request merged**